lib/auth/periodic.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package auth implements certificate signing authority and access control server
+// Authority server is composed of several parts:
+//
+// * Authority server itself that implements signing and acl logic
+// * HTTP server wrapper for authority server
+// * HTTP client wrapper
+package auth
+
+import (
+	"fmt"
+
+	"golang.org/x/mod/semver"
+
+	"github.com/gravitational/teleport/api/types"
+	vc "github.com/gravitational/teleport/lib/versioncontrol"
+)
+
+// upgradeEnrollPeriodic is a periodic operation that aggregates per-version counts of instances
+// by wether or not they are enrolled in automatic upgrades and generates a prompt to enroll
+// instances if the median unenrolled version is lagging behind the median enrolled version.
+type upgradeEnrollPeriodic struct {
+	// enrolled/unenrolled per-version counts
+	enrolled, unenrolled map[string]int
+}
+
+func newUpgradeEnrollPeriodic() *upgradeEnrollPeriodic {
+	return &upgradeEnrollPeriodic{
+		enrolled:   make(map[string]int),
+		unenrolled: make(map[string]int),
+	}
+}
+
+// VisitInstance adds an instance to ongoing aggregations.
+func (u *upgradeEnrollPeriodic) VisitInstance(instance types.Instance) {
+	ver := vc.Normalize(instance.GetTeleportVersion())
+	if !semver.IsValid(ver) {
+		return
+	}
+
+	if instance.GetExternalUpgrader() == "" {
+		u.unenrolled[ver]++
+	} else {
+		u.enrolled[ver]++
+	}
+}
+
+// GenerateEnrollPrompt generates a prompt suggesting enrollment of unenrolled instances
+// in automatic upgrades if the median enrolled version is higher than the median unenrolled version.
+func (u *upgradeEnrollPeriodic) GenerateEnrollPrompt() (msg string, prompt bool) {
+	medianEnrolled, totalEnrolled, ok := inspectVersionCounts(u.enrolled)
+	if !ok || totalEnrolled == 0 {
+		return "", false
+	}
+
+	medianUnenrolled, totalUnenrolled, ok := inspectVersionCounts(u.unenrolled)
+	if !ok || totalUnenrolled == 0 {
+		return "", false
+	}
+
+	if semver.Compare(medianEnrolled, medianUnenrolled) != 1 {
+		// unenrolled agents are not lagging behind enrolled agents
+		return "", false
+	}
+
+	return fmt.Sprintf("some agents are outdated and would benefit from enrollement in automatic upgrades."+
```

Comment:
```suggestion
	return fmt.Sprintf("Some agents are outdated and would benefit from enrollment in automatic upgrades."+
```

Should we capitalize this since it ends in a period, or does this get concatenated to some other message?